diff --git a/webform_simplenews.inc b/webform_simplenews.inc
index 856db54..bed62d9 100644
--- a/webform_simplenews.inc
+++ b/webform_simplenews.inc
@@ -188,38 +188,37 @@ function _webform_submit_newsletter_email($component, $value) {
       }
     }
   }
-  $news_vid = $component['extra']['news_vid'];
 
   if (count($selected_subscriptions) > 0) {
     $news_vid = $selected_subscriptions;
   }
   else {
-    $news_vid = array();
+    $news_vid = $component['extra']['news_vid'];
   }
 
   $action = $component['extra']['action'];
   $account = simplenews_load_user_by_mail($mail);
+
   // If email belongs to the current registered user, don't send confirmation.
   // Other addresses receive a confirmation if double opt-in is selected.
   if ($account->uid && $account->uid == $user->uid) {
     $confirm = FALSE;
   }
-  else {
-    $confirm = variable_get('simplenews_opt_inout_' . $news_vid, 'double') == 'double';
-  }
 
   if ($mail && isset($news_vid)) {
     if ($action == 'subscribe') {
       if (count($news_vid) > 0) {
         foreach ($news_vid as $k => $v) {
-          simplenews_subscribe_user($mail, $v, $confirm);
+          $newsletter = db_query('SELECT * FROM {simplenews_category} sc WHERE sc.tid = :tid', array(':tid' => $v))->fetchObject();
+          simplenews_subscribe_user($mail, $v, $newsletter->opt_inout == 'double');
         }
       }
     }
     else {
       if (count($news_vid) > 0) {
         foreach ($news_vid as $k => $v) {
-          simplenews_unsubscribe_user($mail, $v, $confirm);
+          $newsletter = db_query('SELECT * FROM {simplenews_category} sc WHERE sc.tid = :tid', array(':tid' => $v))->fetchObject();
+          simplenews_unsubscribe_user($mail, $v, $newsletter->opt_inout == 'double');
         }
       }
     }
